posts/react-compound-components/2021-5-25-React-compound-compnents.md

```diff
@@ -0,0 +1,250 @@
+    className
+  ].join(" ");
+
+  return <div className={panelStyles}>{children}</div>;
+};
+``` 
+   
+We have done with creating all other components. Let's see what we have done.
+
+- In the Accordion-item, label prop is used to differentiate between different accordion items. The label prop will be required in the accordionButton and accordionPanel components, so we are adding label prop to the accordionItem children along with the other props.
+- We are using the **useAccordionContext** in the AccordionPanel and AccordionButton. That is how we get the data that is being provided from the Accordion component.
+- We use changeSelectedItem in the AccordionButton component to update the active item when the button is clicked.
+- We use activeItem in the AccordionPanel component whether show the content or hide the content
+
+Now we had built the accordion component completely, let's see how we can use the Accordion component
+
+```tsx
+import {
+  Accordion,
+  AccordionButton,
+  AccordionItem,
+  AccordionPanel
+} from "./components/accordion";
+
+export default function App() {
+  return (
+    <div className="App">
+      <Accordion>
+        <AccordionItem label="react">
+          <AccordionButton>React</AccordionButton>
+          <AccordionPanel>
+            <p>
+              Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do
+              eiusmod tempor incididunt ut labore et dolore magna aliqua.
+            </p>
+          </AccordionPanel>
+        </AccordionItem>
+        <AccordionItem label="angular">
+          <AccordionButton>Angular</AccordionButton>
+          <AccordionPanel>
+            <p>
+              Excepteur sint occaecat cupidatat non proident, sunt in culpa qui
+              officia deserunt mollit anim id est laborum.
+            </p>
+          </AccordionPanel>
+        </AccordionItem>
+        <AccordionItem label="javascipt">
+          <AccordionButton>Javasciprt</AccordionButton>
+          <AccordionPanel>
+            Duis aute irure dolor in reprehenderit in voluptate velit esse
+            cillum dolore eu fugiat nulla pariatur.
+          </AccordionPanel>
+        </AccordionItem>
+      </Accordion>
+    </div>
+  );
+}
+``` 
+Yayyy!!! We had built the accordion component using the compound components.
+
+ We can also build the same accordion component using render props method but there are many limitations to style the inner components (AccordionButton & AccordionPanel) , we need to pass props like renderAccordionButton, buttonClassName for the AccordionButton and we need separate props for AccordionItem and also AccordionPanel. 
+ 
+ Look at the accordion component now, it looks clean, you can style each and every component of Accordion using its respective component. In future, if you want to have buttonColour for the AccordionButton component, you can just add that prop to the AccordionButton alone and not to the outer accordion component.
```

Comment:
The accordion component, now looks clean. One can style every component of the Accordion using its respective component. In future, if you want to have buttonColour for the AccordionButton component, you can just add that prop to the AccordionButton itself. Not clutter the outer accordion component.